packages/cli-commands/src/index.flow.js

```diff
@@ -0,0 +1,15 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ * @format
+ * @oncall react_native
+ */
+
+export {bundleCommand} from './commands/bundle/bundle';
```

Comment:
Can you also export `buildBundleWithConfig`? We're using it in rnx-kit.

Reply:
Looking into this! 👍🏻

Reply:
Before resolving this issue, can you repeat the reasoning behind why you're not going to export `buildBundleWithConfig` for future reference?

Reply:
From offline discussion with @tido64:

**API changes to this package**

We're reducing the surface area of `@react-native/cli-commands` (formerly `@react-native-community/cli-plugin-metro`) to interface as a React Native CLI commands plugin (with no other JS exports). Longer term, it may become a completely internal part of React Native.

This means no longer exporting `buildBundleWithConfig` as a shareable component (previously depended on by `rnx-kit`).

We're doing this because, combined with Metro, the React Native team is already maintaining two stable public APIs for creating a bundle:
- The `react-native bundle` command (this package) — CLI entry point to creating a bundle, configured with a CLI config file + Metro config.
- Metro's `Server` API — JS interface which allows implementers (e.g. Expo, rnx-kit) to [bootstrap their own `bundle` logic](https://github.com/microsoft/rnx-kit/blob/6d05950be69959a0c9b6bb82b4d0ec0d50200b4c/packages/metro-service/src/bundle.ts#L59-L112).

Maintaining `buildBundleWithConfig` as a third function is unnecessary surface area for us. We believe it's reasonable instead for alternative CLIs/tooling to reimplement this logic themselves (using the above Metro APIs), despite some foregone code reuse today.

Reply:
Just to clarify, I did _not_ agree to this removal. I am pretty sure I made this crystal clear.

`cli-plugin-metro` was [created by us](https://github.com/react-native-community/cli/pull/1447) so that we can share the bundle logic between `@react-native-community/cli` and ourselves, and contribute to a greater good. My personal opinion is that us copying all the logic almost verbatim in our own `@rnx-kit/metro-service` is just an overhead for us while providing no gains whatsoever. I also argued that it splits the community because any improvements we make no longer automatically accrues value to anyone not using our CLI and vice versa. If the Metro team insists on this direction, we have no other choice but to go our separate paths.

Reply:
Hey @tido64 - I'm sure we all can come to an understanding here. The structure of dependencies between React Native, Metro and `@react-native-community/cli` is a longstanding pain point that's reduced our ability (as the React Native team at Meta) to improve the developer experience for the benefit of the entire ecosystem. Our aim in rearranging the packages is to facilitate *greater* sharing and more consistency across the major RN integration toolkits (Meta, RN CLI, rnx-kit, Expo). We have **no intention** to "split the community" or force you to do work you don't think is useful. Let's follow up in our upcoming Metro contributors meeting and make sure we have an agreed way forward.

Reply:
Following up from our meeting yesterday, I have a proposal for how to proceed.

## RN 0.73

Since we weren't aware of the `rnx-kit` call site when planning this change, I'd like to first ensure we're not forcing a sudden breaking change on `rnx-kit` without an agreed migration path. To that end, we'll export `unstable_buildBundleWithConfig` from the new `@react-native/community-cli-plugin` package in React Native 0.73.0, with the understanding that **it will likely move or otherwise change** in future versions ( = some time in the next few months). We'll discuss changes with @tido64 ahead of time (but will explicitly *not* support any new call sites / packages beyond existing usage).

## Migration path (RN 0.74+)

For ease of maintenance and deduplicating package responsibilities, `@react-native/community-cli-plugin` will eventually be *purely* a [plugin](https://github.com/react-native-community/cli/blob/main/docs/plugins.md) for `@react-native-community/cli`. That is, it will not have secondary exports. Any other useful functionality it exposes on top of being a plugin will need to move elsewhere, where it can be consumed without being tied specifically to the RN CLI. As such, `unstable_buildBundleWithConfig` will need to be removed, and `rnx-kit` will need an alternative.

My understanding of what `buildBundleWithConfig` does (and how we might split it up) is:

1. It [validates the `platform` argument against the Metro config](https://github.com/react-native-community/cli/blob/716555851b442a83a1bf5e0db27b6226318c9a69/packages/cli-plugin-metro/src/commands/bundle/buildBundle.ts#L54-L70), which can be moved directly into Metro IMO.
2. It [mutates the current environment](https://github.com/react-native-community/cli/blob/716555851b442a83a1bf5e0db27b6226318c9a69/packages/cli-plugin-metro/src/commands/bundle/buildBundle.ts#L74) according to the `dev` parameter, which tbh is a strange side effect I wouldn't want to include in a reusable function. `rnx-kit` should feel free to copy this line of code if it makes sense for its users, or we can just delete it.
3. It [strips path information from `sourceMapUrl`](https://github.com/react-native-community/cli/blob/716555851b442a83a1bf5e0db27b6226318c9a69/packages/cli-plugin-metro/src/commands/bundle/buildBundle.ts#L76-L79) unless `--sourcemap-use-absolute-path` is specified. Not sure about this one, but we might be able to bake this into Metro and remove support for `--sourcemap-use-absolute-path` (I doubt anyone's using this option).
4. It uses documented, typed Metro APIs to [build a bundle and get the asset information for that bundle](https://github.com/react-native-community/cli/blob/716555851b442a83a1bf5e0db27b6226318c9a69/packages/cli-plugin-metro/src/commands/bundle/buildBundle.ts#L81-L101) - this code is trivial and safe to replicate.
5. It [writes out asset files in a platform-dependent way](https://github.com/react-native-community/cli/blob/716555851b442a83a1bf5e0db27b6226318c9a69/packages/cli-plugin-metro/src/commands/bundle/buildBundle.ts#L104-L109). This part is responsible for the bulk of additional logic that `rnx-kit` had to fork from the CLI before `buildBundleWithConfig` existed. I'll discuss this below.

Re: (5) specifically - the build logic for assets should have a single source of truth in React Native, close to the [`@react-native/assets-registry` package](https://github.com/facebook/react-native/tree/main/packages/assets) which *already* implicitly depends on it. It is clearly useful to all apps that use `@react-native/assets-registry`, not just ones that happen to be built with the RN CLI. It might belong in a new package, or *directly* in `@react-native/assets-registry` for that matter (but not in `@react-native/community-cli-plugin`).

@tido64, please let me know whether this seems workable to you. (I can imagine a few alternative designs, but let's start here.) I would love your help in getting the right abstractions in place so that `rnx-kit` continues to be a valuable tool for RN users as we iterate on the core (cross-framework) developer experience.

Reply:
For (5) specifically, the current logic there is currently too platform specific, and only for the built in platforms.  This logic needs to be customizable for additional platforms.  @react-native/assets-registry is one part of this logic, but it does not handle things like writing iOS/macOS assetCatalogs.  And similar to some of the custom file location logic for android, windows will also need some custom file location logic.

I created https://github.com/react-native-community/cli/pull/2002 (which after this PR will need porting to core instead), which would allow platforms to provide their own specific logic for how to save assets collected during the bundle process.